src/Operation/FindOneAndReplace.php

```diff
@@ -160,12 +160,12 @@ public function __construct(string $databaseName, string $collectionName, $filte
 
         if (isset($options['codec'])) {
             $replacement = $options['codec']->encodeIfSupported($replacement);
-        }
 
-        // Psalm's assert-if-true annotation does not work with unions, so
-        // assert the type manually instead of using is_document
-        // See https://github.com/vimeo/psalm/issues/6831
-        assert(is_array($replacement) || is_object($replacement));
+            // Psalm's assert-if-true annotation does not work with unions, so
+            // assert the type manually instead of using is_document
+            // See https://github.com/vimeo/psalm/issues/6831
+            assert(is_array($replacement) || is_object($replacement));
+        }
```

Comment:
Similar to my comments in BulkWrite, I think it makes sense to address this earlier in the constructor _before_ the existing `$replacement` validation. I'm assuming that only the _result_ of the encoding is relevant. Hypothetically, if the application defined an encoder that somehow stripped out update operators from a document and yielded a valid replacement document, that should be valid.

I'll note that this class is already ensuring that encoding happens in the constructor, so we don't have the same issue as BulkWrite with that being deferred to execution time.